plugins/techdocs-addons-test-utils/src/test-utils.tsx

```diff
@@ -0,0 +1,240 @@
+ * limitations under the License.
+ */
+
+import React, { ReactElement, Fragment } from 'react';
+
+// Shadow DOM support for the simple and complete DOM testing utilities
+// https://github.com/testing-library/dom-testing-library/issues/742#issuecomment-674987855
+import { screen } from 'testing-library__dom';
+import { renderToStaticMarkup } from 'react-dom/server';
+import { Route } from 'react-router-dom';
+import { act, render } from '@testing-library/react';
+
+import {
+  wrapInTestApp,
+  TestApiProvider,
+  TestApiProviderProps,
+} from '@backstage/test-utils';
+import { FlatRoutes } from '@backstage/core-app-api';
+
+import {
+  TechDocsAddons,
+  techdocsApiRef,
+  TechDocsEntityMetadata,
+  TechDocsMetadata,
+  techdocsStorageApiRef,
+} from '@backstage/plugin-techdocs-react';
+import { TechDocsReaderPage, techdocsPlugin } from '@backstage/plugin-techdocs';
+import { catalogPlugin } from '@backstage/plugin-catalog';
+import { searchApiRef } from '@backstage/plugin-search-react';
+
+/**
+ * @public
+ */
+export type RecursivePartial<T> = {
+  [P in keyof T]?: RecursivePartial<T[P]>;
+};
+
+const techdocsApi = {
+  getTechDocsMetadata: jest.fn(),
+  getEntityMetadata: jest.fn(),
+};
+
+const techdocsStorageApi = {
+  getApiOrigin: jest.fn(),
+  getEntityDocs: jest.fn(),
+  syncEntityDocs: jest.fn(),
+};
+
+const searchApi = {
+  query: jest.fn().mockResolvedValue({ results: [] }),
+};
+
+/**
+ * @public
+ */
+export type Apis = TestApiProviderProps<any>['apis'];
+
+type TechDocsAddonBuilderOptions = {
+  dom: ReactElement;
+  entity: RecursivePartial<TechDocsEntityMetadata>;
+  metadata: RecursivePartial<TechDocsMetadata>;
+  componentId: string;
+  apis: Apis;
+  path: string;
+};
+
+const defaultOptions: TechDocsAddonBuilderOptions = {
+  dom: <></>,
+  entity: {},
+  metadata: {},
+  componentId: 'docs',
+  apis: [],
+  path: '',
+};
+
+const defaultMetadata = {
+  site_name: 'Tech Docs',
+  site_description: 'Tech Docs',
+};
+
+const defaultEntity = {
+  kind: 'Component',
+  metadata: { namespace: 'default', name: 'docs' },
+};
+
+const defaultDom = (
+  <html lang="en">
+    <head />
+    <body>
+      <div data-md-component="container">
+        <div data-md-component="navigation" />
+        <div data-md-component="toc" />
+        <div data-md-component="main" />
+      </div>
+    </body>
+  </html>
+);
+
+/**
+ * @public
+ */
+
+export class TechDocsAddonBuilder {
+  private options: TechDocsAddonBuilderOptions = defaultOptions;
+  private addons: ReactElement[];
+
+  static buildAddonsInTechDocs(addons: ReactElement[]) {
+    return new TechDocsAddonBuilder(addons);
+  }
+
+  constructor(addons: ReactElement[]) {
+    this.addons = addons;
+  }
+
+  withApis(apis: Apis) {
+    const refs = apis.map(([ref]) => ref);
+    this.options.apis = this.options.apis
+      .filter(([ref]) => !refs.includes(ref))
+      .concat(apis);
+    return this;
+  }
+
+  withDom(dom: ReactElement) {
+    this.options.dom = dom;
```

Comment:
Thought on the pattern here is that the user might expect these not to mutate, but return a copy of the builder with the additional option set. That'd let you create a base builder than you then use in multiple different tests with some additional options.

Reply:
We've found in practice that instantiating a new builder/tester per each assertion is actually pretty compact/clear; wondering if sharing a base builder/tester that gets copied/modified per assertion would make tests less clear.

Ended up keeping the fluent pattern for now.

Reply:
Pattern overall is fine, just cautious about the expectation on how it works. Thinking stuff like this:

```ts
const myTestedAddon = TechDocsAddonTester.buildAddonsInTechDocs(...)

it('works with API A', async () => {
  await myTestedAddon.withApis(...).renderWithEffects()
  ...
})

it('works with API B', async () => {
  await myTestedAddon.withApis(...).renderWithEffects()
  ...
})
```

It's fluent and all, just atm you'd be mutating the builder which will lead to unexpected results.

The downside of creating copies things is that you can't have standalone statements, but that's a bit ugly either way imo:

```ts
if (...) {
  builder.withApis(...)
}
```

Not feeling super strongly either way, just wanted to highlight it for you to consider